### PR TITLE
Adding TEAL cashflow and synthetic history generation to new NE multiperiod notebook

### PR DESCRIPTION
**Pull request by GabrielSoto-INL**
## Addresses issue:
#86 and #90 

## Summary/Motivation:
Add capability to use TEAL Pyomo cashflow expressions and synthetic history generation from RAVEN tools within the DISPATCHES framework. Currently works by having DISPATCHES installed within a RAVEN conda environment (steps provided within new Jupyter notebook).

## Changes proposed in this PR:
- New multiperiod nuclear case notebook that generates synthetic histories from a provided ARMA model and creates Pyomo cashflow expressions with TEAL
- Two new files with TEAL cash flow generation methods and synthetic history generation methods

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md and COPYRIGHT.md file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by lbianchi-lbl** [comment]
Depending on how much we want the code in this PR to be tested in CI, we'll need to make a detailed plan on how to incorporate the RAVEN/TEAL dependency within the CI infrastructure.

**Comment by ksbeattie** [comment]
This is waiting on #157

**Comment by joshua-cogliati-inl** [comment]
Can dispatches/case_studies/nuclear_case/ARMA_Model/output/arma.pk be generated instead of commited to the repository? (It is large, and probably will tend to get out of date.)

**Comment by GabrielSoto-INL** [comment]
I've removed the arma model from the repository and instead added clearer instructions for the user on how to train/generate the model using the pip-installed `raven_framework` package. Perhaps a test can be added for generating the ARMA model by running the included XML script (`raven_framework ARMA_train.xml`)

File: dispatches/case_studies/nuclear_case/SynHist_integration.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+cwd = os.getcwd()
+proj_dir = path.dirname( path.abspath( path.join(cwd, '../../..') ) )
+raven_dir = path.abspath( path.join(proj_dir, 'raven') )
+sys.path.append( proj_dir )
+sys.path.append( raven_dir )
+sys.path.append( path.abspath( path.join(raven_dir, 'scripts')))
+sys.path.append( path.abspath( path.join(raven_dir, 'plugins')))
+
+from raven.ravenframework.utils import xmlUtils
+from raven.scripts import externalROMloader
```
Comment (joshua-cogliati-inl) [inline dispatches/case_studies/nuclear_case/SynHist_integration.py:33]: For a pip install, ravenframework is the toplevel, so this should be:
```
from ravenframework.utils import xmlUtils
from ravenframework import ROMExternal
```

File: dispatches/case_studies/nuclear_case/SynHist_integration.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+  def buildRunner(self):
+    """
+      Builds synthetic history object from a pickledROM.
+      Currently only ARMA models are supported for pickledROM.
+    """
+    if not path.exists(self.target_file):
+      raise Exception(f"Target file not found at {self.target_file}")
+
+    runner = externalROMloader.ravenROMexternal(self.target_file, raven_dir)
```
Comment (joshua-cogliati-inl) [inline dispatches/case_studies/nuclear_case/SynHist_integration.py:66]: This should be switched over from `externalROMloader.ravenROMexternal` to `ROMExternal.ROMLoader`

**Review by joshua-cogliati-inl — COMMENTED** [review]
Some changes will be needed to work with pip installed raven.

File: dispatches/case_studies/nuclear_case/SynHist_integration.py (hunk 1 above)
Comment (joshua-cogliati-inl) [reply dispatches/case_studies/nuclear_case/SynHist_integration.py]: Fixed :)

File: dispatches/case_studies/nuclear_case/SynHist_integration.py (hunk 2 above)
Comment (joshua-cogliati-inl) [reply dispatches/case_studies/nuclear_case/SynHist_integration.py]: Fixed :)

**Comment by lbianchi-lbl** [comment]
Update from the dev call: there are some remaining steps to be done before these notebooks can be merged. @lbianchi-lbl will help with these after the winter break:

- Now that #157 has been merged, we need to make sure the setup instructions in the notebooks in this PR match the actual installation steps for RAVEN/TEAL (basically, just run `pip install -r requirements-raven.txt` after installing DISPATCHES)
- Some changes might be needed to track updates to the DISPATCHES/`idaes.apps.grid_integration` APIs
- There are Git conflicts that need to be resolved (might be enough to re-run the notebook before committing it)

**Comment by radhakrishnatg** [comment]
@GabrielSoto-INL I updated the Jupyter notebook to use the new version of `MultiPeriodModel` class. While running the notebook, I noticed that `arma.pk` file is missing. Are we allowed to include this file in the repo? If not, do you have a version that can be included? I will push the updated file to my fork and send you the link for you to review it. After finalizing everything, we can decide how to merge this PR with the main. Thank you!

**Comment by radhakrishnatg** [comment]
@GabrielSoto-INL My apologies! I just went through the previous comments and understood why that file is missing. I will run `raven_framework ARMA_train.xml` and see if it generates the file.

**Comment by radhakrishnatg** [comment]
@GabrielSoto-INL I updated the notebook to use the latest version of the code https://github.com/radhakrishnatg/dispatches/tree/teal-integration/dispatches/case_studies/nuclear_case
I just need to discuss a couple of points with you. If you are available, we can discuss them over a Teams call. Thank you!

**Comment by radhakrishnatg** [comment]
@GabrielSoto-INL is currently working on a few changes to the price-taker Jupyter notebook. After finalizing the changes, this PR will be ready for review. But, before we can merge the PR, Issue https://github.com/gmlc-dispatches/sample-data/issues/2 needs to be resolved.

**Comment by radhakrishnatg** [comment]
This PR will be replaced by #177

**Pull request closed without merging**